File: notes/manager.py
import itertools
import shutil
from datetime import datetime
from pathlib import Path
from typing import Tuple

from notes import editors, markdown
from notes.file import NoteFile


class NoteManager:
    def __init__(self, path: Path):
        self.path = path
        self.notes = NoteFile(path)

    def new_entry(self, title: str, timestamp: datetime) -> int:
        """
        add a new entry to notes, and return the line position of the new entry

        :param title str: title for the new entry
        :param timestamp datetime: timestamp for the new entry
        :return int: the line position for the new entry in the notes file
        """

        lines = [markdown.format_entry(title, timestamp)]

        # check if todays date is already in the file
        if not self._find_entry_on(timestamp)[1]:
            lines.insert(0, "\n" + markdown.format_date(timestamp) + "\n")

        position = self._find_latest_position(timestamp)
        self.backup()
        return self.notes.write("\n".join(lines) + "\n", position)

    def open_latest_entry(self, editor: str, timestamp: datetime):
        """
        open notes to the latest entry position

        :param editor str: the editor to use for opening the notes
        :param timestamp datetime: timestamp to get the latest entry relative to it
        """
        position = self._find_latest_position(timestamp)
        self.open(editor, position)

    def open(self, editor: str, position: int):
        """
        open notes in an editor at the specified position

        :param editor str: the editor to use for opening the notes
        :param position int: the position to open the editor at
        """
        editors.open(editor, str(self.path), position)

    def backup(self) -> Path:
        """backup the notes file"""
        backup_dir = self.path.parent / ".backups"
        backup_dir.mkdir(exist_ok=True)
        backup_path = backup_dir / datetime.utcnow().strftime("%Y%m%dT%H%M%S.txt")

        if self.path.exists():
            shutil.copy(self.path, backup_path)

        return backup_path

    def _find_latest_position(self, timestamp: datetime) -> int:
        """
        try to look for the entry from the day before, and append the new entry two lines above it
        if no entry was found (e.g. after the weekend or a vacation) find the latest entry instead

        There is an edge case when there is only one entry in the file for today (e.g. first day of the month)
        so there is no entry for yesterday and an entry from today, then the position should be at the end of the file
        i don't expect to see this case anytime, if it happens i might add some work around
        """
        position, found = self._find_entry_before(timestamp)
        if found:
            return position - 1 if position >= 1 else 0

        position, found = self._find_entry_on(timestamp)
        if found and self._has_only_one_entry():
            return len(self.notes)

        else:
            # Dunno how this could happen but 🤷‍♀️
            return position

    def _find_entry_on(self, timestamp: datetime) -> Tuple[int, bool]:
        """find the position of a specific timestamp"""
        formatted = markdown.format_date(timestamp)
        return self.notes.find(lambda line: formatted in line)

    def _find_entry_before(self, timestamp: datetime) -> Tuple[int, bool]:
        """find the first entry before """
        formatted = markdown.format_date(timestamp)
        return self.notes.find(lambda line: markdown.DATE_PATTERN.match(line) and formatted not in line)

    def _has_only_one_entry(self) -> bool:
        entries = itertools.islice(self.notes.find_all(lambda line: markdown.DATE_PATTERN.match(line)), 0, 2)
        return len(list(entries)) == 1

File: notes/manager.go
package notes

import (
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/ahodieb/papyrus/editor"
)

type Manager struct {
	Notes  NotesFile
	Editor editor.EditorOpener
}

// Open notes file in the editor at the end of the latest time entry
func (m *Manager) OpenLatest() error {
	return m.Editor.Open(m.Notes.Path, m.findLatest(time.Now()))
}

// Open opens notes file in the editor at the specified line index
func (m *Manager) Open(i int) error {
	return m.Editor.Open(m.Notes.Path, i)
}

// AddEntry to notes and return its position
// FIXME still not working properly
// TODO restructure
func (m *Manager) AddEntry(title string, t time.Time) (position int) {
	if _, found := m.findOn(t); !found {
		latest, _ := m.findBefore(t)
		latest = m.Notes.Insert([]string{formatDate(t), ""}, latest)
		position = m.Notes.Insert([]string{formatEntry(title, t)}, latest)
	} else {
		position = m.Notes.Insert([]string{formatEntry(title, t)}, m.findLatest(t)+1)
	}

	if position < len(m.Notes.Lines) && m.Notes.Lines[position] != "" {
		m.Notes.Insert([]string{""}, position)
	}

	if _, err := m.Notes.SaveWithBackup(); err != nil {
		fmt.Fprintf(os.Stderr, "Backup failed: %s", err)
	}

	return
}

// findLatest finds the latest entry in the notes
func (m *Manager) findLatest(t time.Time) int {
	// Looks for the latest entry before the current day entry, and get the position two lines above it
	if position, found := m.findBefore(t); found {
		if position > 2 {
			return position - 2
		}

		return 0
	}

	// If no later entry found look for the current day entry,
	// This could happen if there is only one entry in the file, or change of formats
	// If none were found that means either its an empty file or formats are not recognized
	// and it will default back to the 0th position
	return len(m.Notes.Lines) - 1
}

// findOn finds section created on the specified date
// returns the index and a found bool
func (m *Manager) findOn(t time.Time) (int, bool) {
	return m.Notes.FindContains(formatDate(t))
}

func (m *Manager) findBefore(t time.Time) (int, bool) {
	date := t.Truncate(time.Hour * 24)
	for _, s := range m.Sections() {
		if s.Time.Before(date) {
			return s.Index, true
		}
	}

	return 0, false
}

type Section struct {
	Index   int
	Time    time.Time
	Content []string
}

//FIXME It does not account for other sections in the journal file
// Also i want to think about the terminology (section, entry, ...etc)
func (m *Manager) Sections() []Section {
	var sections []Section

	for i, line := range m.Notes.Lines {
		if DATE_PATTERN.MatchString(line) {
			t, err := parseDate(line)
			if err == nil {
				sections = append(sections, Section{Index: i, Time: t})
			}
		}
	}

	for i, section := range sections {
		if i+1 >= len(sections) {
			sections[i].Content = m.Notes.Lines[section.Index:]
		} else {
			sections[i].Content = m.Notes.Lines[section.Index:sections[i+1].Index]
		}
	}

	sort.Slice(sections, func(i, j int) bool {
		return sections[i].Time.After(sections[j].Time)
	})
	return sections
}

var DATE_PATTERN, _ = regexp.Compile("### [a-zA-Z]{3} (?P<year>[0-9]{4})/(?P<month>[0-9]{2})/(?P<day>[0-9]{2})")

const ROUND_DURATION = 5 * time.Minute

func formatDate(t time.Time) string {
	return t.Format("### Mon 2006/01/02")
}

func parseDate(s string) (time.Time, error) {
	return time.Parse("### Mon 2006/01/02", strings.TrimSpace(s))
}

func formatEntry(title string, start time.Time) string {
	formatedTitle := reverseHashTags(title)
	return fmt.Sprintf("* %s | %s/", formatedTitle, floorTime(start, ROUND_DURATION).Format("15:04"))
}

func floorTime(t time.Time, d time.Duration) time.Time {
	round := t.Round(d)
	if round.After(t) {
		return round.Add(-d)
	}
	return round
}

// reverse hashtags in entry title e.g  something# ->  #something
// creating a new entry starting with #something does not work in bash without escaping it
// as a convince i reverse them something# which works fine on bash
func reverseHashTags(s string) string {
	w := strings.Split(s, " ")

	for i := range w {
		if strings.HasSuffix(w[i], "#") {
			w[i] = "#" + strings.TrimSuffix(w[i], "#")
		}
	}

	return strings.Join(w, " ")
}
